pkg/internal/infrastructure/helper.go

```diff
@@ -0,0 +1,122 @@
+)
+
+const (
+	TenantNetworkAnnotation = "tenant.network.infra.gardener.io"
+)
+
+// CreateNetworkAttachmentDefinitionIfNotExist returns the GetNeworkAttachmentDefinition object matching the given GetNeworkAttachmentDefinitionRef.
+// If no existing NetworkAttachmentDefinition found, create and return one using CNI configs.
+func CreateNetworkAttachmentDefinitionIfNotExist(ctx context.Context, c client.Client, secretRef corev1.SecretReference, nadRef []*kubevirtapi.NetworkAttachmentDefinitionRef) ([]networkv1.NetworkAttachmentDefinition, error) {
+	if len(nadRef) == 0 {
+		return nil, nil
+	}
+	netClient, err := getNetworkClient(ctx, c, secretRef)
+	if err != nil {
+		return nil, err
+	}
+	nads := make([]networkv1.NetworkAttachmentDefinition, len(nadRef))
+	for i, ref := range nadRef {
+		// there could be two types of net-attach-def:
+		// pre-created ones for shared resources, e.g. storage network, that each tenant needs to attach
+		// or dynamically created ones on per tenant basis for network isolation.
+		// for pre-created network, simple get and use it. For tenant network, create it using the CNI config, and tenant network annotation.
+		nad, err := netClient.NetworkAttachmentDefinitions(ref.NetAttachDefNamespace).Get(ref.NetAttachDefName, metav1.GetOptions{})
+		if k8serrors.IsAlreadyExists(err) {
+			nads[i] = *nad
+		} else {
+			// if net-attach-def doesn't exist, create one
+			if ref.Config != nil {
+				if nad, err := netClient.NetworkAttachmentDefinitions(ref.NetAttachDefNamespace).Create(
+					&networkv1.NetworkAttachmentDefinition{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: networkName(*ref.Config),
+							Annotations: map[string]string{
+								TenantNetworkAnnotation: "true",
+							},
+						},
+						Spec: networkv1.NetworkAttachmentDefinitionSpec{
+							Config: *ref.Config,
+						},
+					}); err != nil {
+					nads[i] = *nad
+				}
+			}
+		}
+	}
+	return nads, nil
+}
+
+// DeleteTenantNetworkIfExist deletes NetworkAttachmentDefinition annotated as tenant network.
+func DeleteTenantNetworkIfExist(ctx context.Context, c client.Client, secretRef corev1.SecretReference, nads []networkv1.NetworkAttachmentDefinition) error {
+	if len(nads) == 0 {
+		return nil
+	}
+	netClient, err := getNetworkClient(ctx, c, secretRef)
+	if err != nil {
+		return err
+	}
+	for _, nad := range nads {
+		if tenant, ok := nad.ObjectMeta.Annotations[TenantNetworkAnnotation]; ok && tenant == "true" {
+			deleteErr := netClient.NetworkAttachmentDefinitions(nad.ObjectMeta.Namespace).Delete(nad.ObjectMeta.Name, &metav1.DeleteOptions{})
+			if deleteErr != nil && !k8serrors.IsNotFound(deleteErr) {
+				err = deleteErr
+			}
+		}
+	}
+	return err
+}
+
+func networkName(input string) string {
+	in := []byte(input)
+	// see if the config json can be compacted
+	compact := new(bytes.Buffer)
+	err := json.Compact(compact, []byte(input))
+	if err == nil {
+		in = compact.Bytes()
+	}
+	id := utils.EncodeSHA1(in)
+	return "shoot-net-" + id
```

Comment:
The network name should contain the shoot name (or the shoot namespace name) to make sure that NADs belonging to a shoot are recognizable by humans. Also, the SHA could be shortened to e.g. 7 chars.